futures-core/src/task/mod.rs

```diff
@@ -1,35 +1,142 @@
+//! Tasks used to drive a future computation
+//!
+//! It's intended over time a particular operation (such as servicing an HTTP
+//! request) will involve many futures. This entire operation, however, can be
+//! thought of as one unit, as the entire result is essentially just moving
+//! through one large state machine.
+//!
+//! A "task" is the unit of abstraction for what is driving this state machine
+//! and tree of futures forward. A task is used to poll futures and schedule
+//! futures with, and has utilities for sharing data between tasks and handles
+//! for notifying when a future is ready. Each task also has its own set of
+//! task-local data generated by `task_local!`.
+//!
+//! Note that libraries typically should not manage tasks themselves, but rather
+//! leave that to event loops and other "executors" (see the `executor` module),
+//! or by using the `wait` method to create and execute a task directly on the
+//! current thread.
+//!
+//! More information about the task model can be found [online at tokio.rs].
+//!
+//! [online at tokio.rs]: https://tokio.rs/docs/going-deeper-futures/futures-model/
+//!
+//! ## Functions
+//!
+//! There is an important bare function in this module: `current`. The
+//! `current` function returns a handle to the currently running task, panicking
+//! if one isn't present. This handle is then used to later notify the task that
+//! it's ready to make progress through the `Task::notify` method.
+
 use core::fmt;
 use core::marker::PhantomData;
 
-use {Poll, Future, Stream};
-
 mod atomic_task;
 pub use self::atomic_task::AtomicTask;
 
-mod core;
+/// The context for the currently running task.
+pub struct Context<'a> {
+    handle_factory: &'a (Fn() -> NotifyHandle + 'a),
+    id: usize,
+    map: &'a LocalMap,
```

Comment:
Why is this an immutable reference, rather than a mutable one? We are mutating the `LocalMap`, and it should be shared with anyone else at the time, should it?

Reply:
Similarly, is there a reason the `handle_factory` is an `&'a Fn()` instead of an `&'a mut FnMut`?

Reply:
An excellent point! Just a carryover from before, I'll change it to a `mut` everywhere.